File: antia_numerical_methods/c/fun/linrn.c
/*	Linear interpolation in N dimensions

	N : (input) Number of variables
	XB : (input) Array of length N containing the coordinates
		of the point at which interpolation is required
	X : (input) Array of length NXD*N containing the abscissas
	F : (input) Array of dimension exactly F[NDIM[N-1]]...[NDIM[0]]
		containing the function values
		F[IN_1]...[I1][I0]=f(X[0][I0],X[1][I1],...,X[N-1][IN_1])
	NP : (input) Integer array of length N containing the number of tabular points
		NP[I] is the number of tabular point along Ith dimension
	FB : (output) The interpolated value
	NDIM : (input) Integer array of length N, containing the
		dimension of F as declared in calling function
	NXD : (input) Second dimension of array X, NXD >= max(NP[i])
		
	Error status is returned by the value of the function LINRN.
		0 value implies successful execution
		207 implies NP[I]>NDIM[I] or NP[I]<2 for some I

	Required functions : LOCATE
*/

#include <math.h>
#include <stdlib.h>

int locate(double xb, double x[], int np);

int linrn(int n, double xb[], double x[], double f[], int np[], double *fb,
	int ndim[], int nxd)

{
	int i,j,k, nd, low, ndp, index;
	double term;
	int *jn;
	double *hn;

	nd=0;
	jn=(int *) calloc((size_t) (2*n), sizeof(int));
	hn=(double *) calloc((size_t) n, sizeof(double));

/*	Locate the point within a hypercube in the table */

	for(i=0; i<n; ++i) {
		if(np[i]>ndim[i] || np[i]>nxd || np[i]<2) {free(hn); free(jn); return 207;}

		low=locate(xb[i],&x[nd],np[i]);
		jn[i+n]=low;
		hn[i]=(xb[i]-x[low+nd])/(x[low+nd+1]-x[low+nd]);
		nd=nd+nxd;
		jn[i]=0;
	}

	*fb=0.0;
	do {
		index=jn[0]+jn[n];
		ndp=ndim[0];
		for(i=1; i<n; ++i) {
			index=index+(jn[i]+jn[i+n])*ndp;
			ndp=ndp*ndim[i];
		}
	
/*	F(IN[0]+IN[N], IN[1]+IN[1+N], ... ,IN[N-1]+IN[2N-1]) */
		term=f[index];

		for(i=0; i<n; ++i) {
			if(jn[i] == 0) term=term*(1-hn[i]);
			else term=term*hn[i];
		}

		*fb = *fb + term;

/*	select the next point */

		k=0;
		while(k <= n-1) {
			if(jn[k] >= 1) {
				jn[k]=0;
				k=k+1;
			}
			else {
				jn[k]=jn[k]+1;
				break;
			}
		}
	} while(k<n);

	free(hn); free(jn);
	return 0;
}

File: projects/antia_numerical_methods@Beliavsky/f77/sub4/linrn.f
!	Linear interpolation in N dimensions
!
!	N : (input) Number of variables
!	XB : (input) Real array of length N containing the coordinates
!		of the point at which interpolation is required
!	X : (input) Real array of length NXD*N containing the abscissas
!	F : (input) Real array of dimension exactly F(NDIM(1),...,NDIM(N))
!		containing the function values
!		F(I1,I2,...,IN)=f(X(I1,1),X(I2,2),...,X(IN,N))
!	NP : (input) Integer array of length N containing the number of tabular points
!		NP(I) is the number of tabular point along Ith dimension
!	FB : (output) Real variable giving the interpolated value
!	NDIM : (input) Integer array of length N, containing the
!		dimension of F as declared in calling program
!	NXD : (input) First dimension of array X, NXD .GE. Max(NP(i))
!	IN : Integer array of length 2N used as scratch space
!	HN : Real array of length N used as scratch space
!	IER : (output) Error parameter, IER=0 implies successful execution
!		IER=207 implies NP(I)>NDIM(I) or NP(I)<2 for some I
!
!	Required routines : LOCATE

      SUBROUTINE LINRN(N,XB,X,F,NP,FB,NDIM,NXD,IN,HN,IER)
!      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION X(*),XB(N),F(*),NP(N),IN(2*N),HN(N),NDIM(N)

      IER=0
      ND=0
!	Locate the point within a hypercube in the table
      DO 2000 I=1,N
        IF(NP(I).GT.NDIM(I).OR.NP(I).GT.NXD.OR.NP(I).LT.2) THEN
          IER=207
          RETURN
        ENDIF

        LOW=LOCATE(XB(I),X(ND+1),NP(I))
        IN(I+N)=LOW
        HN(I)=(XB(I)-X(LOW+ND))/(X(LOW+ND+1)-X(LOW+ND))
        ND=ND+NXD
        IN(I)=0
2000  CONTINUE

      FB=0.0
3000  INDEX=IN(1)+IN(1+N)
      NDP=NDIM(1)
      DO 3200 I=2,N
        INDEX=INDEX+(IN(I)+IN(I+N)-1)*NDP
        NDP=NDP*NDIM(I)
3200  CONTINUE
!	F(IN(1)+IN(1+N), IN(2)+IN(2+N), ... ,IN(N)+IN(2N))
      TERM=F(INDEX)

      DO 3400 I=1,N
        IF(IN(I).EQ.0) THEN
          TERM=TERM*(1.-HN(I))
        ELSE
          TERM=TERM*HN(I)
        ENDIF
3400  CONTINUE

      FB=FB+TERM

!	select the next point
      K=1
4200  IF(IN(K).GE.1) GO TO 4400
      IN(K)=IN(K)+1
      GO TO 3000

!	If the Kth dimension is exhausted, go to the next one
4400  IN(K)=0
      K=K+1
      IF(K.LE.N) GO TO 4200

      END
